code/game/objects/structures/fires.dm

```diff
@@ -0,0 +1,325 @@
+		return TRUE
+
+	if(ATOM_IS_OPEN_CONTAINER(thing))
+		if(thing.reagents?.total_volume)
+			user.visible_message(SPAN_DANGER("\The [user] pours the contents of \the [thing] into \the [src]!"))
+			take_reagents(thing.reagents)
+			return TRUE
+
+	if(lit == FIRE_LIT)
+
+		if(istype(thing, /obj/item/flame/match))
+			var/obj/item/flame/match/match = thing
+			if(!match.burnt && !match.lit)
+				match.lit = TRUE
+				match.update_icon()
+				return TRUE
+
+		if(istype(thing, /obj/item/flame/candle))
+			var/obj/item/flame/candle/flame = thing
+			flame.light(user)
+			return TRUE
+
+	else if(thing.isflamesource())
+		visible_message(SPAN_NOTICE("\The [user] attempts to light \the [src] with \the [thing]..."))
+		light()
+		return TRUE
+
+	if((lit != FIRE_LIT || user.a_intent == I_HURT) && user.unEquip(thing, src))
+		user.visible_message(SPAN_NOTICE("\The [user] drops \the [thing] into \the [src]."))
+		update_icon()
+		return TRUE
+
+	return ..()
+
+/obj/structure/fire_source/proc/process_fuel()
+
+	if(fuel >= IDEAL_FUEL)
+		return TRUE
+
+	var/list/waste = list()
+	for(var/obj/item/thing in contents)
+
+		if(fuel >= IDEAL_FUEL)
+			break
+
+		if(istype(thing, /obj/item/stack))
+			var/obj/item/stack/stack = thing
+			if(stack.material.fuel_value > 0)
+				var/fuel_per_unit = 2 * stack.material.fuel_value
+				var/use_stacks = min(stack.amount, Floor((IDEAL_FUEL - fuel) / fuel_per_unit))
+				var/add_fuel = round(fuel_per_unit * use_stacks)
+				if(stack.material.burn_product)
+					if(waste[stack.material.burn_product])
+						waste[stack.material.burn_product] += add_fuel
+					else
+						waste[stack.material.burn_product] = add_fuel
+				fuel += add_fuel
+				stack.use(use_stacks)
+				continue
+
+		var/modified_fuel = FALSE
+		for(var/mat in thing.matter)
+			var/decl/material/material = decls_repository.get_decl(mat)
+			if(material.fuel_value > 0)
+				modified_fuel = TRUE
+				var/add_fuel = round(thing.matter[mat] / SHEET_MATERIAL_AMOUNT) * material.fuel_value
+				if(material.burn_product)
+					if(waste[material.burn_product])
+						waste[material.burn_product] += add_fuel
+					else
+						waste[material.burn_product] = add_fuel
+				fuel += add_fuel
+
+		if(modified_fuel)
+			qdel(thing)
+			continue
+
+	// Dump waste gas from burned fuel.
+	var/turf/T = get_turf(src)
+	var/datum/gas_mixture/environment = T?.return_air()
+	if(environment && length(waste))
+		for(var/w in waste)
+			if(waste[w] > 0)
+				environment.adjust_gas(w, waste[w], FALSE)
+		environment.update_values()
+
+	return (fuel > 0)
+
+/obj/structure/fire_source/proc/take_reagents(datum/reagents/RG)
+	var/do_steam = FALSE
+	for(var/rtype in RG.reagent_volumes)
+		var/decl/material/R = decls_repository.get_decl(rtype)
+		if(R.fuel_value <= 0)
+			do_steam = TRUE
+		fuel += REAGENT_VOLUME(RG, rtype) * R.fuel_value
+	RG.clear_reagents()
+	fuel = max(0, fuel)
+	if(lit == FIRE_LIT)
+		if(fuel <= 0)
+			die()
+		if(do_steam)
+			steam.start() // HISSSSSS!
```

Comment:
Does this work OK if called more than once? Does it return steam to original state later? Does it ever try to `qdel` steam, which will fail?

I know there are often deletion problems with these effects because of how they are implemented.

Reply:
Steam overrides start() and just sets up a bunch of callbacks to a proc that creates an effect and moves it around, looks fine to call repeatedly to me.